### PR TITLE
WoodFisher Recipe Tutorial

### PR DESCRIPTION
**Pull request by bfineran**

### CODE REVIEW

File: integrations/pytorch/tutorials/woodfisher_pruning.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+(gradients) to approximate the Inverse-Hessian matrix that is used to score weights
+for pruning.
+
+This method was shown to achieve SoTA results for pruning tasks on
+image classification models such as ResNet and MobileNet. Detailed information of the
+methods and results can be found in the [paper](https://arxiv.org/pdf/2004.14340.pdf).
+
+WoodFisher pruning can be enabled by a single modifier in a SparseML recipe. This tutorial
+will demonstrate how to adapt an existing SparseML pruning recipe to do this.
+
```
Comment (jeanniefinks) [inline integrations/pytorch/tutorials/woodfisher_pruning.md:37]:
```suggestion

## Need Help?
For Neural Magic Support, sign up or log in to get help with your questions in our **Tutorials channel:** [Discourse Forum](https://discuss.neuralmagic.com/) and/or [Slack](https://join.slack.com/t/discuss-neuralmagic/shared_invite/zt-q1a1cnvo-YBoICSIw3L1dmQpjBeDurQ). 

```

File: integrations/pytorch/tutorials/woodfisher_pruning.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+epoch range, sparsity levels, and schedule, so these values may remain
+unchaged when adjusting a recipe.
+
+The `MFACPruningModifier` implements a global pruning scheme however, so all
+parameters to be pruned should be listed under the `params` argument. If the previous
+recipe has multiple pruning modifiers, all of their prameters should be merged into a
+single `MFACPruningModifier`.
+
+WoodFisher specific hyperparameters can be specified under the `mfac_options` parameter.
+Values should be listed as a YAML dictionary.  Full detials about these hyperparameters
```
Comment (jeanniefinks) [inline integrations/pytorch/tutorials/woodfisher_pruning.md:51]:
```suggestion
Values should be listed as a YAML dictionary.  Full details about these hyperparameters
```

File: integrations/pytorch/tutorials/woodfisher_pruning.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+will demonstrate how to adapt an existing SparseML pruning recipe to do this.
+
+## WoodFisher Pruning Recipes
+To use the WoodFisher algorithm in a SparseML pruning recipe, the `MFACPruningModifier`
+must be used in place of the existing pruning modifier(s).  the `MFACPruningModifier`
+contains the same parameters as SparseML's magnitude pruning modifiers for controlling
+epoch range, sparsity levels, and schedule, so these values may remain
+unchaged when adjusting a recipe.
+
+The `MFACPruningModifier` implements a global pruning scheme however, so all
```
Comment (jeanniefinks) [inline integrations/pytorch/tutorials/woodfisher_pruning.md:45]:
```suggestion
The `MFACPruningModifier` implements a global pruning scheme, however, so all
```

File: integrations/pytorch/tutorials/woodfisher_pruning.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+single `MFACPruningModifier`.
+
+WoodFisher specific hyperparameters can be specified under the `mfac_options` parameter.
+Values should be listed as a YAML dictionary.  Full detials about these hyperparameters
+are listed in the next section of this tutorial.
+
+### Example Recipe Section
+
+The following is an example `MFACPruningModifier` to be used in place of other
+PruningModifiers in a recipe:
```
Comment (jeanniefinks) [inline integrations/pytorch/tutorials/woodfisher_pruning.md:57]: PruningModifiers is one string here right?

File: integrations/pytorch/tutorials/woodfisher_pruning.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+
+### fisher_block_size
+To limit the computational cost of calculating second order information, the WoodFisher
+algorithm may compute a block diagonal matrix of a certain block size that is
+sufficient for generating the necessary information for pruning.
+
+The `fisher_block_size` specifies this block size.  If using GPUs to perform the
+WoodFisher computations, the GPUs should have `num_grads * fisher_block_size` extra
+memory duing training so each block can be stored and computed sequentially on a GPU.
```
Comment (jeanniefinks) [inline integrations/pytorch/tutorials/woodfisher_pruning.md:109]:
```suggestion
memory during training so each block can be stored and computed sequentially on a GPU.
```

File: integrations/pytorch/tutorials/woodfisher_pruning.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+with. If not provided, computation will be done on the CPU.
+
+
+## Wrap-Up
+This tutorial shows how to use WoodFisher pruning by swapping in a single modifier
+to a SparseML pruning recipe. Everything else for pruning and exporting stays the same.
+More papers and results will be posted as they become available.
+
+Open an
+[issue](https://github.com/neuralmagic/sparseml/issues) with any questions or bugs.
```
Comment (jeanniefinks) [inline integrations/pytorch/tutorials/woodfisher_pruning.md:126]:
```suggestion

For Neural Magic Support, sign up or log in to get help with your questions in our **Tutorials channel:** [Discourse Forum](https://discuss.neuralmagic.com/) and/or [Slack](https://join.slack.com/t/discuss-neuralmagic/shared_invite/zt-q1a1cnvo-YBoICSIw3L1dmQpjBeDurQ). 

```

Comment (jeanniefinks) [inline integrations/pytorch/tutorials/woodfisher_pruning.md:126]:
```suggestion
[issue](https://github.com/neuralmagic/sparseml/issues) with any bugs or feature requests.

```suggestion

For Neural Magic Support, sign up or log in to get help with your questions in our **Tutorials channel:** [Discourse Forum](https://discuss.neuralmagic.com/) and/or [Slack](https://join.slack.com/t/discuss-neuralmagic/shared_invite/zt-q1a1cnvo-YBoICSIw3L1dmQpjBeDurQ). 

```
```

**Review by jeanniefinks — CHANGES_REQUESTED** [review]
Looks great Benjamin. Just had a few edits!

File: integrations/pytorch/tutorials/woodfisher_pruning.md (hunk 1 above)
Comment (bfineran) [reply integrations/pytorch/tutorials/woodfisher_pruning.md:37]: added to bottom

File: integrations/pytorch/tutorials/woodfisher_pruning.md (hunk 4 above)
Comment (bfineran) [reply integrations/pytorch/tutorials/woodfisher_pruning.md:57]: changing to two lower case strings for consistency. It may be one in the future

**Comment by bfineran** [comment]
thanks for the comments @jeanniefinks, updated!

File: integrations/pytorch/tutorials/woodfisher_pruning.md (hunk 1 above)
Comment (jeanniefinks) [reply integrations/pytorch/tutorials/woodfisher_pruning.md:37]: For the template, we're intentionally upfront letting people know that if they need any help before they actually start the tutorial, there's a method to do it. We then remind them at the end as a wrap-up, to not forget about the offer to help.

Comment (bfineran) [reply integrations/pytorch/tutorials/woodfisher_pruning.md:37]: moved it up

**Comment by bfineran** [comment]
closing in favor of #358

**Pull request closed without merging**